subproviders/subscriptions.js

```diff
@@ -0,0 +1,131 @@
+
+// a cheap crack at multiple inheritance
+// I don't really care if `instanceof EventEmitter` passes...
+Object.assign(SubscriptionSubprovider.prototype, EventEmitter.prototype)
+
+// preserve our constructor, though
+SubscriptionSubprovider.prototype.constructor = SubscriptionSubprovider
+
+SubscriptionSubprovider.prototype.subscribe = function(payload, cb) {
+  const self = this
+  let method = () => {}
+
+  switch (payload.params[ 0 ]) {
+    case 'logs':
+      let options = payload.params[ 1 ]
+
+      method = self.newLogFilter.bind(self, options)
+      break
+    case 'newPendingTransactions':
+      method = self.newPendingTransactionFilter
+      break
+    case 'newHeads':
+      method = self.newBlockFilter
+      cb(null, id)
+      return
+    case 'syncing':
+    default:
+      cb(new Error('unsupported subscription type'))
+      return
+  }
+
+  method(function(err, id) {
+    if (!err) {
+      self.subscriptions[id] = payload.params[0]
+      self.filters[id].on('data', function(results) {
+        if (!Array.isArray(results)) {
+          result = [results]
+        }
+
+        var notificationHandler = self._notificationHandler.bind(self, id, payload.params[0])
+        results.forEach(notificationHandler)
+        self.filters[id].clearChanges()
+      })
+    }
+    cb(err, id)
+  })
+}
+
+SubscriptionSubprovider.prototype._notificationHandler = function (id, subscriptionType, result) {
+  const self = this
+  if (subscriptionType === 'newHeads') {
+    result = {
+      hash: utils.bufferToHex(result.hash),
+      parentHash: utils.bufferToHex(result.parentHash),
+      sha3Uncles: utils.bufferToHex(result.sha3Uncles),
+      miner: utils.bufferToHex(result.miner),
+      stateRoot: utils.bufferToHex(result.stateRoot),
+      transactionsRoot: utils.bufferToHex(result.transactionsRoot),
+      receiptsRoot: utils.bufferToHex(result.receiptsRoot),
+      logsBloom: utils.bufferToHex(result.logsBloom),
+      difficulty: from.intToQuantityHex(utils.bufferToInt(result.difficulty)),
+      number: from.intToQuantityHex(utils.bufferToInt(result.number)),
+      gasLimit: from.intToQuantityHex(utils.bufferToInt(result.gasLimit)),
+      gasUsed: from.intToQuantityHex(utils.bufferToInt(result.gasUsed)),
+      nonce: result.nonce ? utils.bufferToHex(result.nonce): null,
+      timestamp: from.intToQuantityHex(utils.bufferToInt(result.timestamp)),
+      extraData: utils.bufferToHex(result.extraData)
+    }
+  }
+  // it seems that web3 doesn't expect there to be a separate error event
+  // so we must emit null along with the result object
+  self.emit('data', null, {
+    jsonrpc: "2.0",
+    method: "eth_subscription",
+    params: {
+      subscription: id,
+      result: result,
+    },
+  })
+}
+
+SubscriptionSubprovider.prototype.unsubscribe = function(connection, payload, cb) {
+  let subscriptionId = payload.params[0]
+  if (!this.subscriptions.hasOwnProperty(subscriptionId.toLowerCase())) {
```

Comment:
`this.subscriptions[subscriptionId]` should be fine, eh?

Reply:
the id is a hex string - I wanted it to be case insensitive

Reply:
but ew, I forgot to write it in that way over in `subscribe` - I'll fix it there

Reply:
alternatively I could normalize it so that `0x0001` matches `0x1`, etc.

Reply:
ok, forget all that noise - I'm going to do it your way :-D